apps/rush-lib/src/logic/taskRunner/TaskCollection.ts

```diff
@@ -115,16 +115,24 @@ export class TaskCollection {
   /**
    * Checks for projects that indirectly depend on themselves.
    */
-  private _checkForCyclicDependencies(tasks: Iterable<ITask>, dependencyChain: string[]): void {
+  private _checkForCyclicDependencies(
+    tasks: Iterable<ITask>,
+    dependencyChain: string[],
+    alreadyCheckedProjects: Set<string>
+  ): void {
     for (const task of tasks) {
       if (dependencyChain.indexOf(task.name) >= 0) {
         throw new Error('A cyclic dependency was encountered:\n'
           + '  ' + [...dependencyChain, task.name].reverse().join('\n  -> ')
           + '\nConsider using the cyclicDependencyProjects option for rush.json.');
       }
-      dependencyChain.push(task.name);
-      this._checkForCyclicDependencies(task.dependents, dependencyChain);
-      dependencyChain.pop();
+
+      if (!alreadyCheckedProjects.has(task.name)) {
+        alreadyCheckedProjects.add(task.name);
```

Comment:
Is this the only perf optimization?  The rest of the changes look like cosmetic cleanup.

Reply:
The changes in TaskSelector ensure that we only crawl a dependency tree if we haven't seen it already.

Reply:
Specifically the `if (!result.has(project)) { ... }` lines in TaskSelector.